File: examples/python/angzarr/protoname.py
"""Proto type name helpers using DESCRIPTOR reflection.

Provides utilities to extract type names from proto messages without
hardcoding strings.

Example::

    from proto import order_pb2
    from protoname import name, type_url

    # Extract short type name
    name(order_pb2.CreateOrder)  # "CreateOrder"

    # Build full type URL
    type_url(order_pb2.CreateOrder)  # "type.examples/examples.CreateOrder"
"""

from typing import Type, Union

from google.protobuf.message import Message

TYPE_URL_PREFIX = "type.examples/examples."


def name(msg_or_cls: Union[Message, Type[Message]]) -> str:
    """Extract the short type name from a proto message or class.

    Args:
        msg_or_cls: Either a proto message instance or message class.

    Returns:
        The short type name (e.g., "CreateOrder").
    """
    if isinstance(msg_or_cls, type):
        return msg_or_cls.DESCRIPTOR.name
    return msg_or_cls.DESCRIPTOR.name


def type_url(msg_or_cls: Union[Message, Type[Message]]) -> str:
    """Build the full type URL for a proto message or class.

    Args:
        msg_or_cls: Either a proto message instance or message class.

    Returns:
        The full type URL (e.g., "type.examples/examples.CreateOrder").
    """
    return TYPE_URL_PREFIX + name(msg_or_cls)

File: examples/go/angzarr/protoname.go
// Package angzarr provides helpers for working with protobuf type names.
package angzarr

import (
	"google.golang.org/protobuf/proto"
)

// TypeURLPrefix is the shared prefix for all proto type URLs in examples.
const TypeURLPrefix = "type.examples/examples."

// Name extracts the short type name from a proto message using reflection.
// Example: Name(&examples.CreateOrder{}) returns "CreateOrder"
func Name(msg proto.Message) string {
	return string(msg.ProtoReflect().Descriptor().Name())
}

// TypeURL builds the full type URL for a proto message.
// Example: TypeURL(&examples.CreateOrder{}) returns "type.examples/examples.CreateOrder"
func TypeURL(msg proto.Message) string {
	return TypeURLPrefix + Name(msg)
}
